openmdao/core/parallel_group.py

```diff
@@ -148,25 +150,76 @@ def _setup_communicators(self, comm, parent_dir):
         # first, just use simple round robin assignment of requested procs
         # until everybody has what they asked for or we run out
         if requested:
-            while assigned < limit:
-                for i, system in enumerate(subsystems):
-                    if max_req_procs[i] is None or \
-                       assigned_procs[i] < max_req_procs[i]:
-                        assigned_procs[i] += 1
-                        assigned += 1
-                        if assigned == limit:
+            if size >= requested: # we have enough for all subsystems
+                while assigned < limit:
+                    for i, system in enumerate(subsystems):
+                        if max_req_procs[i] is None or \
+                           assigned_procs[i] < max_req_procs[i]:
+                            assigned_procs[i] += 1
+                            assigned += 1
+                            if assigned == limit:
+                                break
+
+                # create buckets (one sub per bucket) to be consistent in how
+                # we split procs below
+                buckets = [(n,[i]) for i,n in enumerate(assigned_procs)]
+
+            else: # we don't have enough, so have to group subsystems
```

Comment:
This grouping code must be similar to whats going on with Parallel FD and related to the utility function you wrote for spitting variables up evenly for distrib comps. Can they all be combined into one utility function?